src/Microsoft.Data/PrimitiveDataFrameColumnArithmetic.cs

```diff
@@ -0,0 +1,4138 @@
+       void Modulo(PrimitiveDataFrameColumnContainer<T> column, T scalar);
+       void And(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right);
+       void And(PrimitiveDataFrameColumnContainer<T> column, T scalar);
+       void Or(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right);
+       void Or(PrimitiveDataFrameColumnContainer<T> column, T scalar);
+       void Xor(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right);
+       void Xor(PrimitiveDataFrameColumnContainer<T> column, T scalar);
+       void LeftShift(PrimitiveDataFrameColumnContainer<T> column, int value);
+       void RightShift(PrimitiveDataFrameColumnContainer<T> column, int value);
+       void Equals(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right, PrimitiveDataFrameColumnContainer<bool> ret);
+       void Equals(PrimitiveDataFrameColumnContainer<T> column, T scalar, PrimitiveDataFrameColumnContainer<bool> ret);
+       void NotEquals(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right, PrimitiveDataFrameColumnContainer<bool> ret);
+       void NotEquals(PrimitiveDataFrameColumnContainer<T> column, T scalar, PrimitiveDataFrameColumnContainer<bool> ret);
+       void GreaterThanOrEqual(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right, PrimitiveDataFrameColumnContainer<bool> ret);
+       void GreaterThanOrEqual(PrimitiveDataFrameColumnContainer<T> column, T scalar, PrimitiveDataFrameColumnContainer<bool> ret);
+       void LessThanOrEqual(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right, PrimitiveDataFrameColumnContainer<bool> ret);
+       void LessThanOrEqual(PrimitiveDataFrameColumnContainer<T> column, T scalar, PrimitiveDataFrameColumnContainer<bool> ret);
+       void GreaterThan(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right, PrimitiveDataFrameColumnContainer<bool> ret);
+       void GreaterThan(PrimitiveDataFrameColumnContainer<T> column, T scalar, PrimitiveDataFrameColumnContainer<bool> ret);
+       void LessThan(PrimitiveDataFrameColumnContainer<T> left, PrimitiveDataFrameColumnContainer<T> right, PrimitiveDataFrameColumnContainer<bool> ret);
+       void LessThan(PrimitiveDataFrameColumnContainer<T> column, T scalar, PrimitiveDataFrameColumnContainer<bool> ret);
+    }
+
+    internal static class PrimitiveDataFrameColumnArithmetic<T>
+        where T : struct
+    {
+        public static IPrimitiveDataFrameColumnArithmetic<T> Instance => PrimitiveDataFrameColumnArithmetic.GetArithmetic<T>();
+    }
+
+    internal static class PrimitiveDataFrameColumnArithmetic
+    {
+        public static IPrimitiveDataFrameColumnArithmetic<T> GetArithmetic<T>()
+            where T : struct
+        {
+            if (typeof(T) == typeof(bool))
```

Comment:
why not use a `switch` statement on the type?

Reply:
you could also set a return value then return `return (IPrimitiveDataFrameColumnArithmetic<T>)value`  which will save all these pasted casts

Reply:
Oh, I just took this from some of the Tensor code we have checked in, so I didn't pay too much attention to it. Will consider it for my next patch

Reply:
How would you specify the `case` where a constant value is required, i.e.
```c#
switch (typeof(T))
{
    case typeof(int):
// ...
```
Won't compile (CS0150) (with C# 7.3 and any version SharpLab offers).
Or does pattern matching provide a syntax I'm not aware of?

Reply:
I'm not an expert on this, but my understanding of this pattern was that the JIT was smart enough to eliminate chunks of code. When JIT'ing for `T` == `bool`, it can see that `if (typeof(T) == typeof(bool))` always holds, similarly `else if (typeof(T) == typeof(int))` never holds, so it can completely remove non-`bool` code paths.

Reply:
@eerhardt correct.

The JIT does this by dead code elimination, as `typeof` is a JIT-time-constant, so all other branches will fall away.

Reply:
Interesting.